packages/builders/src/messages/Message.ts

```diff
@@ -0,0 +1,544 @@
+		this.data.message_reference = updater(this.data.message_reference ?? new MessageReferenceBuilder());
+		return this;
+	}
+
+	/**
+	 * Clears the message reference for this message.
+	 */
+	public clearMessageReference(): this {
+		this.data.message_reference = undefined;
+		return this;
+	}
+
+	/**
+	 * Adds components to this message.
+	 *
+	 * @param components - The components to add
+	 */
+	public addComponents(
+		...components: RestOrArray<
+			| ActionRowBuilder
+			| APIActionRowComponent<APIMessageActionRowComponent>
+			| ((builder: ActionRowBuilder) => ActionRowBuilder)
+		>
+	): this {
+		this.data.components ??= [];
+
+		const resolved = normalizeArray(components).map((component) => resolveBuilder(component, ActionRowBuilder));
+		this.data.components.push(...resolved);
+
+		return this;
+	}
+
+	/**
+	 * Removes, replaces, or inserts components for this message.
+	 *
+	 * @remarks
+	 * This method behaves similarly
+	 * to {@link https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Array/splice | Array.prototype.splice()}.
+	 *
+	 * It's useful for modifying and adjusting order of the already-existing components of a message.
+	 * @example
+	 * Remove the first component:
+	 * ```ts
+	 * message.spliceComponents(0, 1);
+	 * ```
+	 * @example
+	 * Remove the first n components:
+	 * ```ts
+	 * const n = 4;
+	 * message.spliceComponents(0, n);
+	 * ```
+	 * @example
+	 * Remove the last component:
+	 * ```ts
+	 * message.spliceComponents(-1, 1);
+	 * ```
+	 * @param start - The index to start at
+	 * @param deleteCount - The amount of components to remove
+	 * @param components - The components to insert
+	 */
+	public spliceComponents(
+		start: number,
+		deleteCount: number,
+		...components: RestOrArray<
+			| ActionRowBuilder
+			| APIActionRowComponent<APIMessageActionRowComponent>
+			| ((builder: ActionRowBuilder) => ActionRowBuilder)
+		>
+	): this {
+		this.data.components ??= [];
+		const resolved = normalizeArray(components).map((component) => resolveBuilder(component, ActionRowBuilder));
+
+		this.data.components.splice(start, deleteCount, ...resolved);
+		return this;
+	}
+
+	/**
+	 * Sets the components of this message.
+	 *
+	 * @param components - The components to set
+	 */
+	public setComponents(
+		...components: RestOrArray<
+			| ActionRowBuilder
+			| APIActionRowComponent<APIMessageActionRowComponent>
+			| ((builder: ActionRowBuilder) => ActionRowBuilder)
+		>
+	): this {
+		this.data.components = normalizeArray(components).map((component) => resolveBuilder(component, ActionRowBuilder));
+		return this;
+	}
+
+	/**
+	 * Sets the sticker ids of this message.
+	 *
+	 * @param stickerIds - The ids of the stickers to set
+	 */
+	public setStickerIds(...stickerIds: RestOrArray<Snowflake>): this {
+		this.data.sticker_ids = normalizeArray(stickerIds) as MessageBuilderData['sticker_ids'];
```

Comment:
Why was this cast needed again?

Reply:
The type in discord-api-types is `[Snowflake, Snowflake, Snowflake] | [Snowflake, Snowflake] | [Snowflake]` and I don't really think I should be taking that as the param type, it's nasty.

Reply:
Maybe we should expand the dtypes type instead, I'd be down for that